Content.Server/Atmos/Piping/Unary/EntitySystems/GasThermoMachineSystem.cs

```diff
@@ -50,34 +50,46 @@ public override void Initialize()
         private void OnThermoMachineUpdated(EntityUid uid, GasThermoMachineComponent thermoMachine, AtmosDeviceUpdateEvent args)
         {
 
-            if (!(_power.IsPowered(uid))
+            if (!(_power.IsPowered(uid) && TryComp(uid, out ApcPowerReceiverComponent? receiver))
                 || !TryComp(uid, out NodeContainerComponent? nodeContainer)
                 || !_nodeContainer.TryGetNode(nodeContainer, thermoMachine.InletName, out PipeNode? inlet))
             {
                 return;
             }
 
-            var airHeatCapacity = _atmosphereSystem.GetHeatCapacity(inlet.Air);
-            var combinedHeatCapacity = airHeatCapacity + thermoMachine.HeatCapacity;
+            float sign = Math.Sign(thermoMachine.Cp); // 1 if heater, -1 if freezer
 
-            var startEnergy = inlet.Air.Temperature * airHeatCapacity;
-
-            if (!MathHelper.CloseTo(combinedHeatCapacity, 0, 0.001f))
+            // Implement hysteresis
+            float hystTemp = thermoMachine.TargetTemperature;
+            if (thermoMachine.HysteresisState)
             {
-                var combinedEnergy = thermoMachine.HeatCapacity * thermoMachine.TargetTemperature + airHeatCapacity * inlet.Air.Temperature;
-                inlet.Air.Temperature = combinedEnergy / combinedHeatCapacity;
+                hystTemp += sign*thermoMachine.TemperatureTolerance;
+            }
 
-                thermoMachine.LastEnergyDelta = inlet.Air.Temperature * airHeatCapacity - startEnergy;
+            // Stop if we've reached our target temperature.
+            if (sign*inlet.Air.Temperature > sign*hystTemp) // inequality flips if freezer
```

Comment:
probably make local vars for the condition members with a proper name

Reply:
What do you mean here? As in `bool shouldHeatOrFreeze = sign*inlet.Air.Temperature > sign*hystTemp`? I think the comment immediately above makes this plenty obvious but I can make this change if desired.